client/app.jsx

```diff
@@ -0,0 +1,123 @@
+            username: null,
+            people: [],
+            currentMessageId: null
+        };
+    },
+    componentWillMount() {
+        const URL = window.location.hostname + ':' + config.port;
+        this._socket = io.connect(URL, {secure: config.websocket.secure});
+    },
+    updateUsername(username) {
+        this.setState({username});
+    },
+    updateCurrentMessageId(messageid) {
+        this.setState({currentMessageId: messageid});
+    },
+    onLoginIntention(username, people) {
+        this.setState({username, people});
+    },
+    removeListeners() {
+        this._socket.off('login-response');
+        this._socket.off('message');
+        this._socket.off('part');
+        this._socket.off('join');
+        this._socket.off('start-typing-respose');
+        this._socket.off('update-typing-messages');
+    },
+    listenOnLoginResponse(onError, onSuccess, onLogin, username) {
+        this._socket.on('login-response', ({success, people, error}) => {
+            if (!success) {
+                onError(error);
+            }
+            else {
+                onSuccess();
+
+                var peopleList = _.chain(people)
+                    .value();
+
+                onLogin(this.state.username, peopleList);
+            }
+        });
+    },
+    listenOnMessage(onMessage) {
+        this._socket.on('message', (data) => {
+            onMessage(data);
+        });
+    },
+    listenOnPart(onPart, deleteTypingMessage) {
+        this._socket.on('part', (username) => {
+            onPart(username);
+            deleteTypingMessage(username);
+        });
+    },
+    listenOnJoin(onJoin) {
+        this._socket.on('join', (username) => {
+            onJoin(username);
+        });
+    },
+    listenOnStartTypingResponse(onStartTypingResponse) {
+        this._socket.on('start-typing-response', (messageid) => {
+            this.setState({currentMessageId: messageid});
+            onStartTypingResponse(messageid);
+        });
+    },
+    listenOnUpdateTypingMessages(onUpdateTypingMessages) {
+        this._socket.on('update-typing-messages', (messagesTyping) => {
+            onUpdateTypingMessages(messagesTyping);
+        });
+    },
+    render() {
+        return (
+            <HashRouter>
+                <Switch>
+                    <Route exact path='/' render={(props) => (
+                        <Ting {...props} username={this.state.username}
+                                         people={this.state.people}
+                                         onLoginIntention={this.onLoginIntention}
+                                         updateUsername={this.updateUsername}
+                                         socket={this._socket}
+                                         currentMessageId={this.state.currentMessageId}
+                                         updateCurrentMessageId={this.updateCurrentMessageId}
+                                         removeListeners={this.removeListeners}
```

Comment:
It is bad practice to expose this method to the child, as it can have undesired side effects.